compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -769,6 +772,7 @@ class Typer extends Namer
       def handlePattern: Tree = {
         val tpt1 = typedTpt
         if (!ctx.isAfterTyper && pt != defn.ImplicitScrutineeTypeRef)
+          checkMatchable(pt, tree.srcPos, pattern = true)
```

Comment:
This check must be performed after we search for a `TypeTest` a few lines below with `tryWithTypeTest`. If a `TypeTest` is found, then it is matchable. The type test synthetization may in turn emit an error saying that the type test cannot be performed. The current check will break the reflection API.

At first glance, the default case of `tryWithTypeTest` seems to be the place where we should do this check.

The following tests can be added to detect this warning that should not happen.
`tests/pos-special/fatal-warnings/type-test-matchable.scala`
```scala
import scala.language.`3.1-migration`
import scala.reflect.TypeTest

trait Foo:
  type X
  type Y <: X
  def x: X
  given TypeTest[X, Y] = ???

object Test:
  def test(foo: Foo): Unit =
    foo.x match
      case x: foo.Y =>
```